go/test/endtoend/mysqlctl/mysqlctl_test.go

```diff
@@ -0,0 +1,158 @@
+
+		// Start topo server
+		err := clusterInstance.StartTopo()
+		if err != nil {
+			return 1
+		}
+
+		if err := clusterInstance.VtctlProcess.CreateKeyspace(keyspaceName); err != nil {
+			return 1
+		}
+
+		initCluster([]string{"0"}, 2)
+
+		// Collect table paths and ports
+		tablets := clusterInstance.Keyspaces[0].Shards[0].Vttablets
+		for _, tablet := range tablets {
+			if tablet.Type == "master" {
+				masterTablet = tablet
+			} else if tablet.Type != "rdonly" {
+				replicaTablet = tablet
+			}
+		}
+
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+func initCluster(shardNames []string, totalTabletsRequired int) {
+	keyspace := cluster.Keyspace{
+		Name: keyspaceName,
+	}
+	for _, shardName := range shardNames {
+		shard := &cluster.Shard{
+			Name: shardName,
+		}
+
+		for i := 0; i < totalTabletsRequired; i++ {
+			// instantiate vttablet object with reserved ports
+			tabletUID := clusterInstance.GetAndReserveTabletUID()
+			tablet := &cluster.Vttablet{
+				TabletUID: tabletUID,
+				HTTPPort:  clusterInstance.GetAndReservePort(),
+				GrpcPort:  clusterInstance.GetAndReservePort(),
+				MySQLPort: clusterInstance.GetAndReservePort(),
+				Alias:     fmt.Sprintf("%s-%010d", clusterInstance.Cell, tabletUID),
+			}
+			if i == 0 { // Make the first one as master
+				tablet.Type = "master"
+			}
+			// Start Mysqlctl process
+			tablet.MysqlctlProcess = *cluster.MysqlCtlProcessInstance(tablet.TabletUID, tablet.MySQLPort, clusterInstance.TmpDirectory)
+			if err := tablet.MysqlctlProcess.Start(); err != nil {
+				return
+			}
+
+			// start vttablet process
+			tablet.VttabletProcess = *cluster.VttabletProcessInstance(tablet.HTTPPort,
+				tablet.GrpcPort,
+				tablet.TabletUID,
+				clusterInstance.Cell,
+				shardName,
+				keyspaceName,
+				clusterInstance.VtctldProcess.Port,
+				tablet.Type,
+				clusterInstance.TopoProcess.Port,
+				clusterInstance.Hostname,
+				clusterInstance.TmpDirectory,
+				clusterInstance.VtTabletExtraArgs,
+				clusterInstance.EnableSemiSync)
+			tablet.Alias = tablet.VttabletProcess.TabletPath
+
+			if _, err := tablet.VttabletProcess.QueryTablet(fmt.Sprintf("create database vt_%s", keyspace.Name), keyspace.Name, false); err != nil {
+				log.Error(err.Error())
+				return
+			}
+
+			shard.Vttablets = append(shard.Vttablets, *tablet)
+		}
+
+		keyspace.Shards = append(keyspace.Shards, *shard)
+	}
+	clusterInstance.Keyspaces = append(clusterInstance.Keyspaces, keyspace)
+}
+
+func TestRestart(t *testing.T) {
+	err := masterTablet.MysqlctlProcess.Stop()
+	assert.Nil(t, err)
+	masterTablet.MysqlctlProcess.CleanupFiles(masterTablet.TabletUID)
+	err = masterTablet.MysqlctlProcess.Start()
+	assert.Nil(t, err)
+}
+
+func TestAutoDetect(t *testing.T) {
+
+	// Start up tablets with an empty MYSQL_FLAVOR, which means auto-detect
+	sqlFlavor := os.Getenv("MYSQL_FLAVOR")
+	os.Setenv("MYSQL_FLAVOR", "")
+
+	// Start vttablet process, should be in NOT_SERVING state as mysqld is not running
```

Comment:
This should be because `InitShardMaster  has not been done yet`

Reply:
removed comment